### PR TITLE
[Design] POStringLocalizer

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This is a design PR for the POStringLocalizer work. Missing from the is the pluralization work (other than just to parse the file) because we haven't decided on what the shape of IPluralizedStringLocalizer is.

### CODE REVIEW

**Comment by hishamco** [comment]
Finally we have it 😄 , great job @ryanbrandenburg I appreciate it.
One thing why we didn't finalize the pluralization stuff after that build this on top of that, as we know PO files support pluralization by its nature

**Comment by hishamco** [comment]
@ryanbrandenburg can you fix the checks failure?

File: src/Microsoft.Extensions.Localization/Internal/POLines/ContextLine.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Text;
+
+namespace Microsoft.Extensions.Localization.Internal.POLines
+{
+    public class ContextLine : TokenLine
+    {
+        public override string Token
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/ContextLine.cs]:
```csharp
public override string Token => "msgctxt ";
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/ContextLine.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            get
+            {
+                return "msgctxt ";
+            }
+        }
+
+        public override Line Parse(string value)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/ContextLine.cs]:
```csharp
public override Line Parse(string value) =>
    new ContextLine
    {
        Value = TrimToken(new StringBuilder(value))
    };
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/FlagLine.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+using System;
+using System.Text;
+using System.Linq;
+
+namespace Microsoft.Extensions.Localization.Internal.POLines
+{
+    public class FlagLine : CommentBase
+    {
+        public override string Token
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/FlagLine.cs]:
```csharp
public override string Token => "#, ";
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/FlagLine.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            get
+            {
+                return "#, ";
+            }
+        }
+
+        public override Line Parse(string value)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/FlagLine.cs]:
```csharp
public override Line Parse(string value) =>
    new FlagLine
    {
        Value = TrimToken(new StringBuilder(value)).ToString().Split(',').ToList()
    };
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/Line.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+                //TODO: Check for values at the end
+            }
+            else
+            {
+                throw new FormatException("Line malformed");
+            }
+        }
+
+        private bool IsQuote(char character)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/Line.cs]:
```csharp
private bool IsQuote(char character) =>
    character == '"' || character == '\'';
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/Line.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            return character == '"' || character == '\'';
+        }
+
+        private StringBuilder Unescape(StringBuilder value)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/Line.cs]:
```csharp
private StringBuilder Unescape(StringBuilder value) =>
    value.Replace("\\\"", "\"").Replace("\\'", "'");
```

File: src/Microsoft.Extensions.Localization/Internal/POLines/LiteralLine.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Text;
+
+namespace Microsoft.Extensions.Localization.Internal.POLines
+{
+    public class LiteralLine : Line
+    {
+        public override Line Parse(string value)
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/POLines/LiteralLine.cs]:
```csharp
public override Line Parse(string value) =>
    new LiteralLine
    {
        Value = TrimQuotes(new StringBuilder(value))
    };
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
@ryanbrandenburg It will be nice if `POLocalizer` implement `IPluralizeStringLocalizer` that I suggest in PR #305 instead of `IStringLocalizer` because it doesn't support pluralization, so that's why you need to do it yourself in each implementation that may support pluralization. My aim to eliminate the tedious and repetitive code that we wrote in such cases.

Second thing the amount of code to parse PO is too much, can we simplify this like what we did in INI configuration as example?

**Comment by hishamco** [comment]
@sebastienros is that mean that `POStringLocalizer` will not be supported?!!

**Comment by sebastienros** [comment]
Fortunately no. It's just that this branch required more changes so it will be started fresh.

**Comment by hishamco** [comment]
I see, furthermore I saw that implementation is complicated little bit, It would be nice if we can simplify parsing like what I have seen in INI Configuration

thanks

**Pull request closed without merging**